### ISSUE TITLE
I18N string best practices vs. design-principles

### ISSUE THREAD

**Issue by aphillips** [w3ctag/design-principles#454]
I18N maintains a set of [best practices](https://www.w3.org/TR/international-specs) (see also [Editor's Copy](https://w3c.github.io/bp-i18n-specdev)). One set of these pertain to the definition of strings.

Design-principles has a set of best practices related to strings that prefer `DOMString` except when one needs `USVString`. This guidance is a little unclear, since there are non-DOM/non-JS/non-HTML specs that would prefer to use e.g. `xsd:string` or a string definition that is close to `USVString` (based on code points). We (I18N) have recently had to go through this exercise with RDF-star and a couple of other specs and this is causing us to revise our best practices.

It would not be helpful if TAG and I18N recommended different things. Our tendency is to prefer a string definition based on scalar value string with an exception for the space where UTF-16/WTF-16 (see #323) are the best practice vs. Design-principles (which is backwards from that). We also want to develop text that explains to specs that touch on UTF-8 based file formats why they want to use `DOMString` in their interfaces.

Note well: we are not disagreeing with the design principles as currently articulated.

I was [actioned](https://github.com/w3c/i18n-actions/issues/50) with making this issue. Please add the `i18n-needs-resolution` label to this issue. (Shouldn't horizontal groups have permission enough to set horizontal review labels on your document repos?)

**Comment by LeaVerou** [comment]
> I was [actioned](https://github.com/w3c/i18n-actions/issues/50) with making this issue. Please add the `i18n-needs-resolution` label to this issue. (Shouldn't horizontal groups have permission enough to set horizontal review labels on your document repos?)

I wish this were possible. Unfortunately GitHub’s permission model around labels is all or nothing. Anyhow, I added the label.

**Comment by annevk** [comment]
(You could create a triage team and give them triage permissions. That's how WHATWG attempts to solve this. That gives people some power over issues, but no write access.)

**Comment by torgo** [comment]
Thanks for this @aphillips - we agree we need to be more clear - @ylafon took an action on today's TAG call to come up with a PR that adds clarity.

**Comment by ylafon** [comment]
@aphillips reading the note in [bp-i18n-specdev](https://w3c.github.io/bp-i18n-specdev/#char_string), I don't find major inconsistencies between the two version, the section here is discouraging more the use of ByteString.
How is it seen as being different?

**Comment by jyasskin** [comment]
I think our "use DOMString unless you have a specific reason not to." is too strong and appears to contradict i18n's "Unless you have a reason not to, use a string definition consistent with USVString." I like their text:

> Use a string definition consistent with [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString) if your specification does not process the internal value of strings and is not required to check for unpaired surrogate code points, or if your specification pertains to the [[DOM](https://www.w3.org/TR/international-specs/#bib-dom)], defines a JavaScript API or data format, or defines strings as opaque values that are not processed.

I feel like @annevk is likely to have the best sense of what's going to go wrong, if anything, if we copy that into the design principles.

**Comment by annevk** [comment]
Is there really any difference between presenting an HTML `title` element to the end user and any other kind of string? If you are going to convert to a scalar value string for algorithm purposes (URL parser, text encoding, I/O) you might as well use USVString upfront (even though it is arguably wrong as it's not the most efficient way to go about this), but otherwise I think the web platform has decided on plain strings, however ugly we may find them.

The only reason to deviate from this would be some kind of "brave new world" scenario, such as Wasm or ML, and even then I would expect a solid debate.

**Comment by aphillips** [comment]
@ylafon asks:
> I don't find major inconsistencies between the two version, the section here is discouraging more the use of ByteString.
How is it seen as being different?

I18N and TAG's recommendations are, unsurprisingly, close to being the same when you peer under the hood and actually read the details. But TAG starts with "use DOMString unless" and I18N starts with "use USVString unless". Careless readers might perceive these as being opposed to one-another.

Note that I18N uses very subtle wording "a string definition _consistent with_ USVString"

@annevk notes:
> If you are going to convert to a scalar value string for algorithm purposes (URL parser, text encoding, I/O) you might as well use USVString upfront (even though it is arguably wrong as it's not the most efficient way to go about this), but otherwise I think the web platform has decided on plain strings, however ugly we may find them.

I tend to agree: most of the time, a string is a bag of code units. And most of the time there is no need to look inside the bag to see if there are surrogates without a dance partner. We do not mean for implementations to actually use a scalar value representation (i.e. the 32-bit one) on the wire _ever_. 

In such a world, the difference is actually about whether we require some code point hygiene or not. It's probably cleaner to use TAG's approach, but our WG members have pushed back on this in the past. 

What we want is something pithy and straightforward, where users do not have to read the attendant documentation in order to know what to do and which will NOT result in different interpretations. We want to avoid:

```idl
dictionary X {
   DOMString foo;
   USVString bar;    // why??
   DOMString baz;
};
```

Perhaps something more like the following (unseen by my colleagues and written here the first time):

> Use DOMString unless you have a specific reason not to, especially when specifying document formats, data structures, protocols, or for API interfaces.
>
> Use USVString when specifying algorithms or processes that handle the contents of strings or for cases in which unpaired surrogates would result in an error.

I have added this discussion to I18N's agenda for our teleconference tomorrow (2024-12-05)

**Comment by aphillips** [comment]
I have just now published an updated version of our best practices doc, with a revised section [here](https://www.w3.org/TR/international-specs/#char_string) dealing with these recommendations. 

At the last moment we decided **not** to include the pithy "Use a DOMString unless you have a reason not to" best practice. It was felt that the "Use a DOMString when.../Use a USVString when..." guidance was sufficient, along with a new best practice of not mixing (similar to my comment just above)

Could @jyasskin @ylafon review this issue again with an eye towards potentially copying/referencing/harmonizing with our best practices? @annevk, would value your comments also.

**Comment by annevk** [comment]
The text around `ByteString` seems wrong. (We should perhaps rename that to `IsomorphicString` as per https://infra.spec.whatwg.org/#isomorphic-string for clarity.) It's quite useful for networking APIs and does not require a particular encoding, the conversion is always the same.

Otherwise the text seems reasonable, but you might want to double check the links for the various terms. Having "surrogate" link to Infra but then "code unit" right after seems bit confusing.

**Comment by aphillips** [comment]
Thanks @annevk. Great callouts.

> The text around ByteString seems wrong. (We should perhaps rename that to IsomorphicString as per https://infra.spec.whatwg.org/#isomorphic-string for clarity.) It's quite useful for networking APIs and does not require a particular encoding, the conversion is always the same.

The goal with the discussion of `ByteString` is to be consistent with CharMod's C011:
> [C011](https://www.w3.org/TR/charmod/#C011)  [S]  Specifications SHOULD NOT define a string as a 'byte string'.

That is, we don't anywhere, I think, want people to use octet sequences to be interpreted as character strings. There is, as you say, a good deal of usefulness in defining things in terms of bytes in networking APIs (and similar) and particularly there is value in equating ASCII bytes with ASCII characters in that domain. And the ability to transmit bytes as a classical string (`char*`) instead of requiring `uint8` arrays shouldn't be overlooked. Our current guidance doesn't capture this particularly well, nor does it mention `IsomorphicString`, which it really should. I'll follow up with an issue/PR.

**Comment by jyasskin** [comment]
@aphillips, is there anything left to do in this repository? If so, are you interested in sending us a PR to do it?

**Comment by aphillips** [comment]
@jyasskin Yes, I think there will be something to do here. I've been caught up in a [substantial rewrite](https://deploy-preview-155--bp-i18n-specdev.netlify.app/#characters) of the section related to characters and strings in our best practices, which is blocking work on Anne's comments (w3c/bp-i18n-specdev#151) from getting landed. As soon as we're done with both (this month) I would be glad to make a PR on your document to ensure harmony.

**Comment by jyasskin** [comment]
Great, thank you!
